packages/browser-runtime-core/src/interpreter/interpreter.ts

```diff
@@ -26,10 +26,12 @@ export class Interpreter {
   }
 
   async evaluate(code: string): Promise<ShellResult> {
-    let result;
+    let result = toShellResult(undefined);
     const contextObjext = this.environment.getContextObject();
 
-    contextObjext[LAST_EXPRESSION_CALLBACK_FUNCTION_NAME] = (val): void => {
+    // TODO(addaleax): Is val actually a Promise for ShellResult or are we just pretending it is?
+    // What if this function here never gets called?
+    contextObjext[LAST_EXPRESSION_CALLBACK_FUNCTION_NAME] = (val: Promise<ShellResult>): void => {
```

Comment:
Still not sure about this one :)

Reply:
`val` doesn't have to be a `Promise`, the fun part is that i don't know what is making that a `ShellResult` either, essentially that function is going to be injected in the code like this:

```
interpreter.evaluate(`someExpr(); 1 + 1`)
=>

iframe.eval(`someExpr(); ___MONGOSH_LAST_EXPRESSION_CALLBACK (1 + 1)`);
```

I don't think the result there is a `ShellResult`

Reply:
Ok, then it’s not just me, and the previous code was just wrong…?

Reply:
I feel sorry about the confusion of this code, i've always thought (i still do :)) that at some point we would have reconciled the whole JS preprocessing (browser friendly adaptations, async-rewrite ...) and JS interpretation layer (IFRAME.eval, vm.runScript, repl eval) and have only one.

I feel the fact that I've used inheritance with the `OpenContextRuntime` deserves an apology on its own. Don't know what got into me, but was completely unnecessary.